src/modules-load/modules-load.c

```diff
@@ -150,34 +267,70 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 log_warning_errno(r, "Failed to parse kernel command line, ignoring: %m");
 
-        r = module_setup_context(&ctx);
+        /* Create a socketpair for communication with probe workers */
+        r = RET_NERRNO(socketpair(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0, pair));
         if (r < 0)
-                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+                return log_error_errno(r, "Failed to create socket pair: %m");
+
+        /* Determine number of workers, either from env or from online CPUs */
+        num_threads = 0;
+
+        const char *e = secure_getenv("SYSTEMD_MODULES_LOAD_NUM_THREADS");
+        if (e) {
+                r = safe_atou(e, &num_threads);
+                if (r < 0)
+                        log_debug_errno(r, "Invalid value in $SYSTEMD_MODULES_LOAD_NUM_THREADS: %s", e);
+        }
+
+        if (num_threads <= 0) {
+                /* By default, use a number of worker threads equal the number of online CPUs,
+                 * but clamp it to avoid a probing storm on machines with many CPUs. */
+                const long ncpus = sysconf(_SC_NPROCESSORS_ONLN);
+                if (ncpus < 0)
+                        log_warning_errno(errno, "Failed to get number of online CPUs: %m");
+                num_threads = CLAMP(ncpus, 1, 16);
+        }
+
+        /* One of the probe threads is the main process */
+        num_threads--;
+
+        r = create_worker_threads(FD_TO_PTR(pair[1]), &threads, &num_threads);
```

Comment:
hmm, this seems a bit overkill: spawning a bazillion of worker threads even if we don't need any seems pointless. I mean, the typical case after all is the one where we have no kmods to load manually like this, because all will be auto-loaded via udev. By spawning one worker thread per CPU here you do a *lot* of work completely unnecessarily.

Hence, please spawn another worker thread for each module we actually want to load, until the limit is reached. i.e. that would mean with 4 cpus, but only 2 kmods configured to be loaded manually, you'd get only 2 worker threads. But if you have 8 kmods configure to be loaded manually but 4 cpus, it would cap out at 4.

Reply:
Cap added.